rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,258 @@
+
+`State` here can be pretty broad, you could also consume context or create effects. This hook should be what the component relies on to function/render
+
+```typescript
+import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utils';
+
+/**
+ * Defines the different slots that can be rendered in this component
+ *
+ * This means that an icon can be rendered 'somewhere'
+ */
+export const sampleShorthandProps = ['icon'];
+
+// Creates a helper function to merge props while respecting slot props
+const mergeProps = makeMergeProps({ deepMerge: sampleShorthandProps });
+
+/**
+ * @parameter props -> these are normal React props for JSX components
+ * @parameter ref -> In case someone wants a ref to the root DOM element
+ * @parameter defaultProps -> safe default props, we set defaults internally but if consumer wants component variants with different defaults, they can set them
+ */
+export const useSample = (props: SampleProps, ref: React.Ref<HTMLElement>, defaultProps?: SampleProps): SampleState => {
+  const resolvedRef = ref || React.useRef();
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
+
+```typescript
+import { makeStyles, ax } from '@fluentui/react-make-styles';
+import { SampleState } from './Sample.types';
+
+/**
+* Styles for the root slot
+*/
+export const useRootStyles = makeStyles([
+  [
+    null,
+    theme => ({
+      backgroundColor: theme.neutralColorTokens.neutralBackground1,
+      paddingRight: '12px',
+    }),
+  ],
+]);
+
+/**
+* Styles for the icon slot, uses state selectors from SampleState
+*/
+export const useIconStyles = makeStyles<Pick<SampleState, 'someState'>>([
+  [
+    // Conditionally apply styles
+    (someState) => someState == 1 ? true : false,
+    () => ({
+      width: '20px',
+      height: '20px',
+      marginRight: '9px',
+    }),
+  ],
+]);
+
+/**
+* Applies style classnames to slots
+*/
+export const useSampleStyles = (state: SampleState) => {
+  const rootClassName = useRootStyles({});
+  const iconClassName = useIconStyles({ someState: state.someState });
```

Comment:
```suggestion
  const rootClassName = useRootStyles(state);
  const iconClassName = useIconStyles(state);
```

Reply:
I would rather pass down the whole `state` object as it costs nothing 🐱

Reply:
should style hooks always use `all` state ?

Reply:
For history: we need this in v0 as we use passed props as a cache key, `makeStyles()` is not using it.

Reply:
Let's chat offline, I think that adding all state might still break encapsulation

Reply:
Might I request that whatever you agree on offline, you'll summarise here please?

Reply:
Decision is to use `state` since it doesn't really make a difference internally since we ship our `useStyles` hooks as is without customization. The only difference being that we would need to write a bunch of utility types for internal use only

If consumers would like to override styles they would need to call `makeStyles` anyway and can type it however they want